src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+    }
+    else if (i % 5 == 3)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(boundaries.colptr(col_b),\
+                                  col_size, row_size).t();
+      ++col_b;
+    }
+    else
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(images.colptr(col_i),
+                                  col_size, row_size).t();
+      ++col_i;  
+    }
+    row_idx += row_size;
+  }
+  return input_data;
+} 
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+GetFeatureDimension()
+{
+  /*
+  shrink: amount to shrink channels
+  p_size: size of image patches
+  n_cell: number of self similarity cells
+  n_orient: number of orientations per gradient scale
+  */
+  arma::vec FtrDim(2);
+
+  const size_t shrink = this->options["shrink"];
+  const size_t p_size = this->options["p_size"];
+  const size_t n_cell = this->options["n_cell"];
+  const size_t rgbd = this->options["rgbd"];
+  const size_t n_orient = this->options["n_orient"];
+  /*
+  n_color_ch: number of color channels
+  n_grad_ch: number of gradient channels
+  n_ch: total number of channels
+  */
+  size_t n_color_ch;
+  if (this->options["rgbd"] == 0)
+    n_color_ch = 3;
+  else
+    n_color_ch = 4;
+
+  const size_t n_grad_ch = 2 * (1 + n_orient);
+
+  const size_t n_ch = n_color_ch + n_grad_ch;
+  FtrDim[0] = std::pow((p_size / shrink) , 2) * n_ch;
+  FtrDim[1] = std::pow(n_cell , 2) * (std::pow (n_cell, 2) - 1) / 2 * n_ch;
+  return FtrDim;
+}
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+DistanceTransform1D(arma::vec const &f, const size_t n, const double inf)
+{
+  arma::vec d(n), v(n), z(n + 1);
+  size_t k = 0;
+  v[0] = 0.0;
+  z[0] = -inf;
+  z[1] = inf;
+  for (size_t q = 1; q <= n - 1; ++q)
+  {
+    float s  = ( (f[q] + q * q)-( f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    while (s <= z[k])
+    {
+      --k;
+      s  = ( (f[q] + q * q) - (f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    }
+
+    ++k;
+    v[k] = static_cast<double>(q);
+    z[k] = s;
+    z[k+1] = inf;
+  }
+
+  k = 0;
+  for (size_t q = 0; q <= n-1; q++)
+  {
+    while (z[k+1] < q)
+      ++k;
+    d[q] = (q - v[k]) * (q - v[k]) + f[v[k]];
+  }
+  return d;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+DistanceTransform2D(MatType &im, const double inf)
+{
+  arma::vec f(std::max(im.n_rows, im.n_cols));
+  // transform along columns
+  for (size_t x = 0; x < im.n_cols; ++x)
+  {
+    f.subvec(0, im.n_rows - 1) = im.col(x);
+    arma::vec d = this->DistanceTransform1D(f, im.n_rows, inf);
+    im.col(x) = d;
+  }
+
+  // transform along rows
+  for (size_t y = 0; y < im.n_rows; y++)
+  {
+    f.subvec(0, im.n_cols - 1) = im.row(y).t();
+    arma::vec d = this->DistanceTransform1D(f, im.n_cols, inf);
+    im.row(y) = d.t();
```

Comment:
What kind of type should we declare at the function--DistanceTransform1D?
arma::vec const &f or the view type of row really return?
I know the first one(arma::vec const &f) can work, problem is are you sure this type will the address
but not the copy of row?I used Eigen before, this library is quite picky about this part, not sure about how armadillo handle it.